### PR TITLE
WIP: arm kexec

### PR DESCRIPTION
**Pull request by rjoleary**
This is very very preliminary. It might compile if you're lucky.

### CODE REVIEW

File: pkg/kexec/kexec_linux.go [hunk 1]
```diff
@@ -1,15 +1,41 @@
...
+	Initramfs *os.File
+
+	// The command line for the next kernel.
+	CmdLine string
+
+	// Set to true to prepend the RSDP to the command line. (amd64 only)
+	PrependRSDP bool
+
+	// Path to the device tree. (arm/arm64 only)
+	DTBPath string
```
Comment (hugelgupf) [inline pkg/kexec/kexec_linux.go]: make this an *os.File like Initramfs? awkward to pass some as files and some as paths

File: pkg/kexec/kexec_linux_arm.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	maxCommandLineSize = 1023 // 1024 with the null character
+	stackSize          = 0x1000
+	heapSize           = 0x10000
+	textOffset         = 0x8000
+)
+
+// FileLoad loads the given kernel as the new kernel with the given options.
+//
+// For arm, this uses the kexec_load(2) syscall.
+func (opts *LinuxOpts) FileLoad(kernel *os.File) error {
```
Comment (hugelgupf) [inline pkg/kexec/kexec_linux_arm.go]: maybe leave these on their own, and look at pkg/boot.LinuxImage? I'm not sure what goes where, just making you aware

File: pkg/kexec/kexec_linux.go (hunk 1 above)
Comment (rminnich) [reply pkg/kexec/kexec_linux.go]: yeah

File: pkg/kexec/kexec_linux.go [hunk 3]
```diff
@@ -1,15 +1,41 @@
...
+	// closing this file. This value is optional.
+	Initramfs *os.File
+
+	// The command line for the next kernel.
+	CmdLine string
+
+	// Set to true to prepend the RSDP to the command line. (amd64 only)
+	PrependRSDP bool
+
+	// Path to the device tree. (arm/arm64 only)
```
Comment (rminnich) [inline pkg/kexec/kexec_linux.go]: honest, I'm convinced, the architecture qualifications are not needed. I also just realized this ought to work on riscv.

File: pkg/kexec/kexec_linux_other.go [hunk 4]
```diff
@@ -1,8 +1,8 @@
-// Copyright 2015-2017 the u-root Authors. All rights reserved
+// Copyright 2015-2019 the u-root Authors. All rights reserved
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build linux,!amd64
+// +build linux,!amd64,!arm
```
Comment (rminnich) [inline pkg/kexec/kexec_linux_other.go]: but if it's named linux_other, and kexec_linux_amd64.go and kexec_linux_arm exist, why do you need this build tag? Or don't you just need the !amd64 and !arm?

File: pkg/kexec/kexec_linux.go (hunk 1 above)
Comment (rminnich) [reply pkg/kexec/kexec_linux.go]: or an io.Reader?

File: pkg/kexec/kexec_linux_other.go (hunk 4 above)
Comment (rjoleary) [reply pkg/kexec/kexec_linux_other.go]: The `other` build tag doesn't have any special meaning to the compiler as far as I'm aware. This line prevents the compiler complaining about multiple definitions of `FileLoad`. It's a catch-all for architectures which aren't amd64 and aren't arm.

Alternatively, we can just delete this file and make it a compile-time error to use this package from an unsupported platform.

File: pkg/kexec/kexec_linux.go (hunk 1 above)
Comment (rjoleary) [reply pkg/kexec/kexec_linux.go]: If this is a `*os.File` and `DefaultLinuxOpts` opens the file, who is responsible for closing it?

File: cmds/kexec/dtb_linux.go [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2019 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build linux,arm linux,arm64
```
[the remark this answers is not in the document]
Comment (rjoleary) [reply cmds/kexec/dtb_linux.go]: Ack, I'll have it for each platform.

Should there be at lease a warning if you specify a device tree when one isn't used?

File: pkg/kexec/kexec_linux.go (hunk 1 above)
Comment (rjoleary) [reply pkg/kexec/kexec_linux.go]: I know -- I'll remove `DefaultLinuxOpts` altogether so the caller explicitly specifies and opens the file.

**Comment by rjoleary** [comment]
This is quite a bit out-of-sync with master. I'm going to rebase and re-open in another PR.

**Pull request closed without merging**